<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Pengetahuan-Dasar-Python" data-toc-modified-id="Pengetahuan-Dasar-Python-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Pengetahuan Dasar Python</a></span><ul class="toc-item"><li><span><a href="#Virtual-Environtment" data-toc-modified-id="Virtual-Environtment-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Virtual Environtment</a></span></li><li><span><a href="#Data-Frame" data-toc-modified-id="Data-Frame-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Data Frame</a></span></li><li><span><a href="#Tipe-tipe-Data" data-toc-modified-id="Tipe-tipe-Data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Tipe-tipe Data</a></span><ul class="toc-item"><li><span><a href="#Quiz-1" data-toc-modified-id="Quiz-1-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Quiz 1</a></span></li></ul></li></ul></li></ul></div>

**LearnPy: Python for Data Analyst**

* Author : Team Algoritma
* Developed by [Algoritma](https://algorit.ma)'s product division and instructors team

# Pengetahuan Dasar Python

Hi *future Data Analyst*!! Pada notebook ini kita akan membahas mengenai analisis data menggunakan bahasa pemrograman Python menggunakan salah satu library yang cukup powerfull yaitu `pandas`. Namun sebelum itu kita akan membahas beberapa hal terlebih dahulu sebelum masuk ke pengenalan kita terhadap `pandas`.

## Virtual Environtment

Virtual environtment merupakan alat yang digunakan untuk menyimpan dependensi dari suatu library atau module yang diperlukan ketika mengerjakan suatu project tertentu. Beberapa project tertentu membutuhkan dependency python yang berbeda, oleh karena itu cukup menyulitkan jika kita membuat virtual environtment. 

Beberapa hal yang perlu dipersiapkan dalam membuat virtual environtment yaitu:
- sudah memiliki Anaconda Navigator/Miniconda
- menggunakan CLI (command line interface) sebagai alat untuk membuat, memanage, dan mengaktifkan virtual environtment

Ketika kita sudah memiliki Anaconda Navigator, biasanya tersedia CLI bawaan dari Anaconda Navigator yaitu Anaconda Prompt. 

Untuk dapat menggunakan library tersebut, kita harus melakukan import library terlebih dahulu. Untuk melakukan import library cukup menggunakan function `import`.

In [2]:
import pandas as pd
print(pd.__version__)

0.23.0


## Data Frame 

Kita akan coba untuk melakukan exploratory data menggunakan data turnover suatu perusahaan. Untuk melihat datanya kita harus mengimport data turnover tersebut yang tersimpan dengan nama file `turnover_balance.csv`. Data ini menunjukkan suatu record history karyawan pada perusahaan tertentu yang mengalami turnover berdasarkan beberapa atribut-atribut penunjangnya yaitu `satisfaction_level`, `last_evaluation`, `number_project`, `average_monthly_hours`, `time_spend_company`, `Work_accident`, `promotion_last_5years`, `division`, dan `salary`. Berikut data turnover tersebut :

In [3]:
turnover = pd.read_csv("data_input/turnover_balance.csv")
turnover.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.82             0.68               3                   140   
1                0.79             0.67               5                   156   
2                0.73             0.95               3                   149   
3                0.92             0.78               3                   218   
4                0.69             1.00               5                   237   

   time_spend_company  Work_accident  promotion_last_5years     division  \
0                   2              0                      0        sales   
1                   2              0                      0  product_mng   
2                   2              0                      0      support   
3                   3              0                      0    technical   
4                   3              0                      0    technical   

  salary  left  
0    low     0  
1    low     0  
2    low     0  
3    low     0  
4   high     0

Berdasarkan data `turnover` diatas, berikut ini adalah glimpse dari setiap variabelnya :

* `satisfaction_level` : tingkat kepuasan karyawan bekerja di suatu perusahaan
* `last_evaluation` : tingkat kepuasan karyawan pada evaluasi terakhir
* `number_project` : banyaknya project yang telah diterima karyawan
* `average_monthly_hours` : rata-rata jam kerja per bulan
* `time_spend_company` : lama waktu bekerja dalam perusahaan (tahun)
* `Work_accident` : ada tidaknya kecelakaan kerja, 0 = tidak ada, 1 = ada
* `promotion_last_5years` : pernah mendapatkan promoksi dalam 5 tahun terakhir, 0 = tidak, 1 = ya
* `division` : nama departemen atau divisi
* `salary` : tingkat pendapatan, dibagi menjadi low, medium dan high
* `left` : data history karyawan resign, 0 = tidak, 1 = ya

Sebelum melangkah pada hal-hal yang rumit, kita akan coba untuk mengulik dari hal yang sederhana terlebih dahulu.

## Tipe-tipe Data 

Sebagai seorang data analyst, kita harus mengetahui terlebih dahulu tipe-tipe data yang akan kita analisis. Pada Python, ketika kita bekerja menggunakan library `pandas` untuk melakukan pengecekan tipe data pada data yang digunakan, kita bisa menggunakan `dtypes`. Nilai yang akan dikeluarkan oleh `dtypes` adalah sebuah series yang berisi tipe data setiap variabel pada suatu data frame.

In [4]:
turnover.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
promotion_last_5years      int64
division                  object
salary                    object
left                       int64
dtype: object

Tipe-tipe data yang terdapat pada Python umumnya adalah:

* **float** : tipe data yang berisi angka kontinu
* **int** : tipe data yang berisi angka bulat
* **object** : tipe data yang berisi alphabetic, seperti nama, alamat, dst
* **bool** : tipe data yang berisi TRUE atau FALSE
* **str** : tipe data yang berisi text

Sebagai data analyst, berdasarkan data yang kita miliki, kita ingin mencoba mencari tahu berapa besar rata-rata jam kerja bulanan. Untuk mengetahui nilai rata-rata pada suatu series bisa menggunakan `.mean()`.

In [5]:
turnover["average_montly_hours"].mean()

203.36614393727248

Berdasarkan data yang ada, diketahui bahwa rata-rata pekerja yang ada yaitu sebesar 203 jam per bulan. Atau jika dikonversikan ke dalam hari yaitu rata-rata jam kerjanya sebesar 10 jam per hari.

In [6]:
turnover["time_spend_company"].mean()

3.6160739288714647

Berdasarkan data yang ada, diketahui bahwa rata-rata lama bekerja dalam suatu perusahaan yaitu 3.6 tahun. 

Dalam beberapa kondisi, tipe data pada suatu kolom atau variabel sangatlah penting untuk diperhatikan. Hal ini akan mempermudah seorang data analyst mengetahui data apa yang sedang dia kerjakan. Selain itu juga ketika akan membuat suatu report, variabel tertentu dapat diinterpretasikan dengan baik menggunakan pengukuran statistik ataupun berupa visualisasi. Oleh karena itu, ketika tipe data dalam data frame ada yang tidak sesuai, kita perlu menyesuaikan agar data dapat direpresentasikan dengan baik. 

Pada course material, telah dibahas bagaimana mengubah tipe data pada suatu data frame. Untuk melakukan perubahan tipe data, kita bisa menggunakan fungsi `astype()` pada kolom yang ingin kita ubah.

**Knowledge check:**

Sebelum kita akan mengalisis kolom manakah yang belum sesuai, kita cek terlebih dahulu tipe data dari masing-masing kolom pada object `turnover` yang kita miliki.

In [ ]:
## your code here


**Latihan!**

Jika kita perhatikan berdasarkan gloosary dari masing-masing kolom pada data `turnover` dan tipe datanya, bisakah Anda temukan kolom mana saja yang harus diubah dalam bentuk `category`? 

Jika sudah menemukannya, Anda dapat mengubahnya menggunakan fungsi `astype('category')` dan jangan lupa simpan kembali agar data `turnover` sudah memiliki tipe data yang tepat untuk masing-masing kolomnya.

In [7]:
## your code here


### Quiz 1

Jika diketahui data sebagai berikut.

In [11]:
toko = pd.DataFrame({
    'jenis' : ['Sepatu', 'Sandal', 'Baju', 'Tas'],
    'warna' : ['Hitam', 'Merah', 'Biru', 'Pink'],
    'harga' : [1300000, 56000, 85000, 100000],
    'diskon' : [0.2, 0, 0, 0.1],
    'stok' : [59, 100, 200, 50],
    'penjualan_mingguan' : [6, 50, 150, 40]
})
toko

jenis  warna    harga  diskon  stok  penjualan_mingguan
0  Sepatu  Hitam  1300000     0.2    59                   6
1  Sandal  Merah    56000     0.0   100                  50
2    Baju   Biru    85000     0.0   200                 150
3     Tas   Pink   100000     0.1    50                  40

1. Jika Fafilia adalah seorang data analyst pada toko tersebuut, dia diminta untuk melakukan analisis terhadap data toko tersebut. Manakah dari kolom yang tersedia termasuk `category` atau `object`?

    - [ ] jenis dan harga
    - [ ] warna dan stok
    - [ ] jenis dan warna
    - [ ] warna dan diskon

2. Manakah kolom yang termasuk `integer`?

    - [ ] diskon
    - [ ] stok
    - [ ] jenis
    - [ ] warna

3. Jika Fafilia diminta untuk melakukan rekap penjualan mingguan dari seluruh barang yang dijual, berapakah rata-rata penjualan mingguannya?

    - [ ] 50
    - [ ] 68.9
    - [ ] 78.5
    - [ ] 61.5

# 